File: lib/features/history/models/ride_model.dart

enum RideStatus {
  pending,
  accepted,
  ongoing,
  completed,
  cancelled
}

class Ride {
  final String id;
  final String pickupLocationName;
  final String dropoffLocationName;
  final List<double> pickupCoordinates;
  final List<double> dropoffCoordinates;
  final double distance;
  final int duration;
  final double fare;
  final String paymentMethod;
  final RideStatus status;
  final DateTime createdAt;
  final String? driverId;
  final DriverInfo? driver;
  final double? rating;
  final String? review;

  Ride({
    required this.id,
    required this.pickupLocationName,
    required this.dropoffLocationName,
    required this.pickupCoordinates,
    required this.dropoffCoordinates,
    required this.distance,
    required this.duration,
    required this.fare,
    required this.paymentMethod,
    required this.status,
    required this.createdAt,
    this.driverId,
    this.driver,
    this.rating,
    this.review,
  });

  factory Ride.fromJson(Map<String, dynamic> json) {
    return Ride(
      id: json['id'],
      pickupLocationName: json['pickupLocation']['name'],
      dropoffLocationName: json['dropoffLocation']['name'],
      pickupCoordinates: List<double>.from(json['pickupLocation']['coordinates'].map((x) => x.toDouble())),
      dropoffCoordinates: List<double>.from(json['dropoffLocation']['coordinates'].map((x) => x.toDouble())),
      distance: json['distance'].toDouble(),
      duration: json['duration'],
      fare: json['fare'].toDouble(),
      paymentMethod: json['paymentMethod'],
      status: _parseStatus(json['status']),
      createdAt: DateTime.parse(json['createdAt']),
      driverId: json['driverId'],
      driver: json['driver'] != null ? DriverInfo.fromJson(json['driver']) : null,
      rating: json['rating']?.toDouble(),
      review: json['review'],
    );
  }

  static RideStatus _parseStatus(String status) {
    switch (status.toLowerCase()) {
      case 'pending':
        return RideStatus.pending;
      case 'accepted':
        return RideStatus.accepted;
      case 'ongoing':
        return RideStatus.ongoing;
      case 'completed':
        return RideStatus.completed;
      case 'cancelled':
        return RideStatus.cancelled;
      default:
        return RideStatus.pending;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'pickupLocation': {
        'name': pickupLocationName,
        'coordinates': pickupCoordinates,
      },
      'dropoffLocation': {
        'name': dropoffLocationName,
        'coordinates': dropoffCoordinates,
      },
      'distance': distance,
      'duration': duration,
      'fare': fare,
      'paymentMethod': paymentMethod,
      'status': status.toString().split('.').last,
      'createdAt': createdAt.toIso8601String(),
      'driverId': driverId,
      'driver': driver?.toJson(),
      'rating': rating,
      'review': review,
    };
  }
}

class DriverInfo {
  final String id;
  final String firstName;
  final String lastName;
  final String? profilePicture;
  final String phone;
  final VehicleInfo vehicle;

  DriverInfo({
    required this.id,
    required this.firstName,
    required this.lastName,
    this.profilePicture,
    required this.phone,
    required this.vehicle,
  });

  String get fullName => '$firstName $lastName';

  factory DriverInfo.fromJson(Map<String, dynamic> json) {
    return DriverInfo(
      id: json['id'],
      firstName: json['firstName'],
      lastName: json['lastName'],
      profilePicture: json['profilePicture'],
      phone: json['phone'],
      vehicle: VehicleInfo.fromJson(json['vehicle']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'firstName': firstName,
      'lastName': lastName,
      'profilePicture': profilePicture,
      'phone': phone,
      'vehicle': vehicle.toJson(),
    };
  }
}

class VehicleInfo {
  final String type;
  final String make;
  final String model;
  final String color;
  final String licensePlate;

  VehicleInfo({
    required this.type,
    required this.make,
    required this.model,
    required this.color,
    required this.licensePlate,
  });

  factory VehicleInfo.fromJson(Map<String, dynamic> json) {
    return VehicleInfo(
      type: json['type'],
      make: json['make'],
      model: json['model'],
      color: json['color'],
      licensePlate: json['licensePlate'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'make': make,
      'model': model,
      'color': color,
      'licensePlate': licensePlate,
    };
  }
}

File: Dash_backend/src/models/rideModel.js
const mongoose = require('mongoose');

const rideSchema = new mongoose.Schema(
  {
    rider: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    driver: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Driver',
      default: null,
    },
    status: {
      type: String,
      enum: ['requested', 'accepted', 'ongoing', 'completed', 'cancelled'],
      default: 'requested',
    },
    pickupLocation: {
      name: {
        type: String,
        required: true,
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: true,
      },
    },
    dropoffLocation: {
      name: {
        type: String,
        required: true,
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: true,
      },
    },
    distance: {
      type: Number, // in kilometers
      required: true,
    },
    duration: {
      type: Number, // in minutes
      required: true,
    },
    fare: {
      type: Number,
      required: true,
    },
    commissionAmount: {
      type: Number,
      default: 0,
    },
    driverAmount: {
      type: Number,
      default: 0,
    },
    paymentMethod: {
      type: String,
      enum: ['cash', 'wallet'],
      required: true,
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'paid'],
      default: 'pending',
    },
    startTime: {
      type: Date,
      default: null,
    },
    endTime: {
      type: Date,
      default: null,
    },
    riderRating: {
      type: Number,
      min: 1,
      max: 5,
      default: null,
    },
    driverRating: {
      type: Number,
      min: 1,
      max: 5,
      default: null,
    },
    cancellationReason: {
      type: String,
      default: null,
    },
    cancelledBy: {
      type: String,
      enum: ['rider', 'driver', 'system', null],
      default: null,
    },
    createdAt: {
      type: Date,
      default: Date.now,
    },
    updatedAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Create index for geospatial queries
rideSchema.index({ 'pickupLocation.coordinates': '2dsphere' });

// Virtual field for rider's full name
rideSchema.virtual('riderName').get(function () {
  if (this.rider && typeof this.rider === 'object') {
    return this.rider.firstName && this.rider.lastName
      ? `${this.rider.firstName} ${this.rider.lastName}`
      : undefined;
  }
  return undefined;
});

// Virtual field for rider's phone number
rideSchema.virtual('riderPhone').get(function () {
  if (this.rider && typeof this.rider === 'object') {
    return this.rider.phone;
  }
  return undefined;
});

// Virtual field to determine if the ride has been rated
rideSchema.virtual('isRated').get(function() {
  // From driver's perspective
  if (this.driver && this.driverRating !== null) {
    return true;
  }
  
  // From rider's perspective
  if (this.rider && this.riderRating !== null) {
    return true;
  }
  
  return false;
});

// Pre-find middleware to populate rider data for virtuals
rideSchema.pre(/^find/, function(next) {
  this.populate({
    path: 'rider',
    select: 'firstName lastName phone'
  });
  next();
});

const Ride = mongoose.model('Ride', rideSchema);

module.exports = Ride;
